core/src/main/java/org/apache/druid/data/input/impl/JsonReader.java

```diff
@@ -33,13 +40,98 @@
 
 import java.io.IOException;
 import java.util.Collections;
-import java.util.List;
+import java.util.Iterator;
 import java.util.Map;
+import java.util.NoSuchElementException;
 
-public class JsonReader extends TextReader
+/**
+ * <pre>
+ * In constract to {@link JsonLineReader} which processes input text line by line independently,
+ * this class tries to parse the input text as a whole to an array of objects.
+ *
+ * The input text can be:
+ * 1. a JSON string of an object in a line or multiple lines(such as pretty-printed JSON text)
+ * 2. multiple JSON object strings concated by white space character(s)
+ *
+ * For case 2, what should be noticed is that if an exception is thrown when parsing one JSON string,
+ * the rest JSON text will all be ignored
+ *
+ * For more information, see: https://github.com/apache/druid/pull/10383
+ * </pre>
+ */
+public class JsonReader implements InputEntityReader
```

Comment:
Thinking about this new reader, I think the new requirement for parsing multiple JSON objects to multiple rows doesn't fit in the current interface of the sampler. The sampler currently assumes that _there is only one JSON object in an input chunk_ which could have either an array or a nested object. That means, the current interface allows multiple out rows as `inputRows` is a list of `InputRow`s in `InputRowListPlusRawValues`, but doesn't allow multiple values in an input chunk as `rawValues` is a map in `InputRowListPlusRawValues`. To support the requirement, I think `rawValues` should be a list of map, so that all raw values in an input chunk can be returned to the sampler.

If we do this, `JsonReader` can simply extend `IntermediateRowParsingReader`, where the intermediate row will be the [`inputText` created from `source.open()`](https://github.com/apache/druid/pull/10383/files#diff-cc591396000acab135e9bd4aa1f31e2aR72). Here is an example.

```java
public class JsonReader2 extends IntermediateRowParsingReader<String>
{
  private final ObjectFlattener<JsonNode> flattener;
  private final ObjectMapper mapper;
  private final InputEntity source;
  private final InputRowSchema inputRowSchema;

  JsonReader2(
      InputRowSchema inputRowSchema,
      InputEntity source,
      JSONPathSpec flattenSpec,
      ObjectMapper mapper,
      boolean keepNullColumns
  )
  {
    this.inputRowSchema = inputRowSchema;
    this.source = source;
    this.flattener = ObjectFlatteners.create(flattenSpec, new JSONFlattenerMaker(keepNullColumns));
    this.mapper = mapper;
  }

  @Override
  protected CloseableIterator<String> intermediateRowIterator() throws IOException
  {
    return CloseableIterators.withEmptyBaggage(
        Iterators.singletonIterator(IOUtils.toString(source.open(), StringUtils.UTF8_STRING))
    );
  }

  @Override
  protected List<InputRow> parseInputRows(String intermediateRow) throws IOException, ParseException
  {
    try (JsonParser parser = new JsonFactory().createParser(intermediateRow)) {
      final Iterator<JsonNode> delegate = mapper.readValues(parser, JsonNode.class);
      return FluentIterable.from(() -> delegate)
                           .transform(jsonNode -> MapInputRowParser.parse(inputRowSchema, flattener.flatten(jsonNode)))
                           .toList();
    }
  }

  @Override
  protected List<Map<String, Object>> toMap(String intermediateRow) throws IOException
  {
    try (JsonParser parser = new JsonFactory().createParser(intermediateRow)) {
      final Iterator<Map> delegate = mapper.readValues(parser, Map.class);
      return FluentIterable.from(() -> delegate).transform(map -> (Map<String, Object>) map).toList();
    }
  }
}
```

Reply:
> The sampler currently assumes that there is only one JSON object in an input chunk which could have either an array or a nested object. 

That's the root cause why `ExceptionThrowingIterator` is extracted and `JsonReader` inherits from InputEntityReader directly.

Your suggestion provides a new and simple way to deal with it. I'll test the code later.